bindings/go/src/examples/perftest/perftest.go

```diff
@@ -0,0 +1,397 @@
+		params.EnableTag()
+	}
+
+	if perfTestParams.wakeup {
+		params.EnableWakeup()
+	}
+
+	perfTest.context, _ = NewUcpContext(params)
+}
+
+func initMemory() error {
+	var err error
+	var dummyMemh *UcpMemory
+	memTypeMask, _ := perfTest.context.MemoryTypesMask()
+
+	if !IsMemTypeSupported(perfTestParams.memType, memTypeMask) {
+		return errors.New("requested memory type is unsupported.")
+	}
+
+	mmapParams := &UcpMmapParams{}
+	// Allocate dummy host memory, to initialize cuda context if the memType is Cuda.
+	mmapParams.SetMemoryType(UCS_MEMORY_TYPE_HOST).Allocate()
+	mmapParams.SetLength(1)
+	dummyMemh, err = perfTest.context.MemMap(mmapParams)
+	if err != nil {
+		return err
+	}
+	dummyMemh.Close()
+
+	mmapParams.SetMemoryType(perfTestParams.memType).Allocate()
+	mmapParams.SetLength(perfTestParams.messageSize * uint64(perfTestParams.numThreads))
+
+	perfTest.memory, err = perfTest.context.MemMap(mmapParams)
+	if err != nil {
+		return err
+	}
+
+	perfTest.memParams, err = perfTest.memory.Query(UCP_MEM_ATTR_FIELD_ADDRESS)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func initWorker(i int) {
+	workerParams := &UcpWorkerParams{}
+
+	if perfTestParams.wakeup {
+		workerParams.WakeupTX()
+		workerParams.WakeupRX()
+	}
+
+	if perfTestParams.numThreads > 1 {
+		workerParams.SetThreadMode(UCS_THREAD_MODE_MULTI)
+	}
+
+	perfTest.perThreadWorkers[i], _ = perfTest.context.NewWorker(workerParams)
+}
+
+func connectWorker(i int) error {
+	var err error
+	epParams := &UcpEpParams{}
+	serverAddress, _ := net.ResolveTCPAddr("tcp", fmt.Sprintf("%v:%v", perfTestParams.ip, perfTestParams.port))
+	epParams.SetPeerErrorHandling().SetErrorHandler(func(ep *UcpEp, status UcsStatus) {
+		if status != UCS_ERR_CONNECTION_RESET {
+			fmt.Printf("Endpoint error: %v \n", status.String())
+		}
+	}).SetSocketAddress(serverAddress)
+
+	perfTest.eps[i], err = perfTest.perThreadWorkers[i].NewEndpoint(epParams)
+
+	if err != nil {
+		return err
+	}
+
+	request, err := perfTest.eps[i].FlushNonBlocking(nil)
+
+	if err != nil {
+		return err
+	}
+
+	for request.GetStatus() == UCS_INPROGRESS {
+		progressWorker(i)
+	}
+
+	if status := request.GetStatus(); status != UCS_OK {
+		return NewUcxError(status)
+	}
+
+	request.Close()
+	return nil
+}
+
+func initListener() error {
+	var err error
+	listenerParams := &UcpListenerParams{}
+	addr, _ := net.ResolveTCPAddr("tcp", fmt.Sprintf("0.0.0.0:%v", perfTestParams.port))
+
+	listenerParams.SetSocketAddress(addr)
+	listenerParams.SetConnectionHandler(func(connRequest *UcpConnectionRequest) {
+		backwardEp, _ := perfTest.perThreadWorkers[0].NewEndpoint((&UcpEpParams{}).
+			SetConnRequest(connRequest).SetPeerErrorHandling())
+
+		// Current connection establishment requires to create backward ep or reject connection.
+		perfTest.backwardEps = append(perfTest.backwardEps, backwardEp)
+		fmt.Println("Got connection starting benchmark")
+	})
+
+	perfTest.listener, err = perfTest.perThreadWorkers[0].NewListener(listenerParams)
```

Comment:
need to have multiple threads on the receiver
the listener will create reply ep on different workers/threads